### PR TITLE
Feature/appointment changing and cancelling

### LINKED ISSUE #263: De gebruiker kan een afspraak annuleren of wijzigen via een url in een bevestigings e-mail.
Eis: Afspraken 3.
Zie ook NvI: 60, 92

**Tasks**

* [x] Add API endpoint to cancel appointment.
* [ ] ~Add API endpoint to change the appointment.~
* [x] The API endpoints should take an appointment ID and an emailaddress as validation. The emailaddress should match the mail address where the confirmation mail is sent.
* [x] Add page (view/template/js/css-whatever) to cancel the appointment (see design below) + confirmation page (Afspraak succesvol geannuleerd).
* [ ] To change an appointment, we can use the link that was used to create the appointment in the first place. So, point the user to the "appointment form", keep track that it was an existing appointment, create the appointment afterwards and cancel the old one.
* [ ] Improvement over the previous bullet: Allow the pre-selection of a product using a query param: `/form/my-form?product=XYZ` and use this in the change appointment link (see below).
* [x] Add links to these pages in the confirmation email: 
  **Afspraak wijzigen of annuleren**
  o [Ik wil mijn afspraak annuleren]()
  o [Ik wil mijn afspraak wijzigen]()

**Design / mock**

image

### PR DESCRIPTION
**Pull request by shea-maykinmedia**
Fixes #263 

~~Relies on https://github.com/open-formulieren/open-forms/pull/573 being merged first~~

**Screenshot**

I'm unsure how accurate this styling is to how the real email will look.  I can do a modification on the styling if this does not look correct.

Screenshot 2021-09-06 at 09 14 23

### CODE REVIEW

File: src/openforms/appointments/base.py
Comment (sergei-maertens) [inline src/openforms/appointments/base.py:234]: _please_ base64 encode the submission ID to not plainly expose database IDs. I'd even say to just base64 the submission UUID instead of using the database PK. See also the django password reset URLs: https://github.com/django/django/blob/3921b1c6d24c9d5a60e5f5f83c9a394104089c21/django/contrib/auth/views.py#L270

File: src/openforms/appointments/base.py
Comment (sergei-maertens) [inline src/openforms/appointments/base.py:239]: Let's decouple getting the links from the rendering in the template.

This method should just return a dict with the links & any templates should call this method instead and use the links.

I want the option to easily include this information in other places, like and API endpoint used by the SDK without being constrained by the template engine.

File: src/openforms/appointments/exceptions.py
Comment (sergei-maertens) [inline src/openforms/appointments/exceptions.py:20]: It doesn't seem like this is a _client_ error, which 4xx indicates. The client did nothing wrong.

I would rather expect HTTP 503 here or perhaps HTTP 502

@joeribekker any preference/input?

File: src/openforms/submissions/utils.py
Comment (sergei-maertens) [inline src/openforms/submissions/utils.py:93]: Let's stop hardcoding these things and just piling on top of content, and instead implement a custom template tag such as `{% appointment_information %}` and `{% get_appointment_links as links %}` so that these can be templated out by the end user.

The end result should be that end end-user can do something like this:

```django
{# other content such as summary #}

{% appointment_information %}

{% get_appointment_links as links %}

U kan deze afspraak <a href="{{ links.cancel_url }}">annuleren</a>
```

& then of course document these options with the rest of the templates documentation.

File: src/openforms/config/models.py
Comment (sergei-maertens) [inline src/openforms/config/models.py:141]: Please rename this specifically to "cancel_appointment_page". I'm also expecting that a global option will not be sufficient and there will be pages specific per form, so this may get a form-specific override in the future. Let's hold off on that for now though. Either way, forms will be embedded in many different CMS pages and there is no "single sdk website".

Help text should also be something like "URL to the page where the appointment cancellation should  happen."

The URL to the SDK is confusing since there's an SDK URL setting which points to which javascript bundle to load.

File: src/openforms/submissions/tokens.py
Comment (sergei-maertens) [inline src/openforms/submissions/tokens.py:107]: Let's put this one in the `appointments` app - since we require tokens that are valid until the appointment date.

So also the `get_expiry_days` based on a setting is wrong - we can have an exact validity of a token based on the appointment date, let's use that information instead of having a token that's valid even if the appointment has passed or not valid long enough.

File: src/openforms/submissions/tokens.py
Comment (sergei-maertens) [inline src/openforms/submissions/tokens.py:188]: * submission report is not relevant for appointments, at all.
* the attributes also do not _change_ after the appointment has changed, and that's the whole premise of a token system like this.

After cancellation, the appointmentinfo is either deleted or updated -> _that_ should be included in the token attributes & other items that may change (appointment date/time comes to mind immediately).

This means that once an appointment is cancelled or modified, the links are now invalidated. See again the Django built in password reset token logic - tokens are invalidated after the password hash has changed there, which makes the link single-use.

File: src/openforms/submissions/tokens.py
Comment (sergei-maertens) [inline src/openforms/submissions/tokens.py:198]: bad variable naming

File: src/openforms/conf/base.py
Comment (sergei-maertens) [inline src/openforms/conf/base.py:459]: this setting should not exist in the context of appointments

File: src/openforms/conf/base.py
Comment (sergei-maertens) [inline src/openforms/conf/base.py:194]: this is a **massive** security risk, I'm hoping it's an accident and not intentional

File: src/openforms/appointments/tokens.py
Comment (sergei-maertens) [inline src/openforms/appointments/tokens.py:7]: Let's implement this entire generator here rather than subclassing - there's too much stuff in motion at the moment that will cause confusion.

File: src/openforms/appointments/tokens.py
Comment (sergei-maertens) [inline src/openforms/appointments/tokens.py:16]: I do not understand the logic here, why should these exceptions not cause problems? Why should it fall back to some weird behaviour? This goes against the Zen of Python "Refuse the temptation to guess"

File: src/openforms/appointments/tests/test_models.py
Comment (sergei-maertens) [inline src/openforms/appointments/tests/test_models.py:6]:
```suggestion
from .factories import AppointmentInfoFactory
```

**Review by sergei-maertens — CHANGES_REQUESTED** [review]
`src/openforms/utils/helpers.py` is an empty file, please remove it

File: src/openforms/conf/base.py
Comment (shea-maykinmedia) [reply src/openforms/conf/base.py:194]: Yes, this was me doing a bit of debugging and accidentally not reverted.

File: src/openforms/appointments/exceptions.py
Comment (shea-maykinmedia) [reply src/openforms/appointments/exceptions.py]: For the moment I've made this a 502.

File: docs/manual/templates.rst
Comment (shea-maykinmedia) [inline docs/manual/templates.rst:127]: Not sure if we want more documentation on this since this essentially adds information that _can_ be used but doesn't display any information itself which makes it different than `appointment_information` or `summary`.

File: src/openforms/appointments/base.py
Comment (shea-maykinmedia) [inline src/openforms/appointments/base.py:243]: Is this the best way to build the full url?  Since we don't have access to the request here I was hoping we can just use this settings variable.

File: src/openforms/appointments/base.py
Comment (shea-maykinmedia) [inline src/openforms/appointments/base.py:240]: Is this the best way to build the full url?  We don't have access to the request so we can't do `request.build_absolute_url()`(or whatever the exact function is) and even if we did I'm not sure that would be the best way to do it?

File: src/openforms/appointments/tests/test_endpoints.py
Comment (shea-maykinmedia) [inline src/openforms/appointments/tests/test_endpoints.py:48]: Is this the correct way to assert this?  At least for this unit test I don't think there should ever be a second submission in the session.

File: src/openforms/appointments/api/tests/test_endpoints_with_jcc.py
Comment (sergei-maertens) [inline src/openforms/appointments/api/tests/test_endpoints_with_jcc.py:339]: See earlier remark about test setup

File: src/openforms/appointments/api/tests/test_endpoints_with_qmatic.py
Comment (sergei-maertens) [inline src/openforms/appointments/api/tests/test_endpoints_with_qmatic.py:265]: See earlier remark about the test setup

File: src/openforms/appointments/api/urls.py
Comment (sergei-maertens) [inline src/openforms/appointments/api/urls.py:17]:
```suggestion
        "<uuid:submission_uuid>/cancel",
```
https://docs.djangoproject.com/en/2.2/topics/http/urls/#path-converters there is a `uuid` converter

File: src/openforms/appointments/api/views.py
Comment (sergei-maertens) [inline src/openforms/appointments/api/views.py:216]: A comment describing the business logic/requirement from the ticket here would provide context that is now completely absent.

File: src/openforms/appointments/base.py
Comment (sergei-maertens) [reply src/openforms/appointments/base.py:240]: Yeah I think it is, for now at least. It does mean that we use the `BASE_URL` setting for more than DigiD, so it should be documented as such.

An alternative is using `django.contrib.sites` but I really dislike that and brings its own problems, so let's improve on documentation here.

However, don't join URLs together like that, as the `BASE_URL` setting may have a trailing slash. To be safe, you should use `urllib.parse.urljoin`.

File: src/openforms/appointments/tests/test_endpoints.py
Comment (sergei-maertens) [reply src/openforms/appointments/tests/test_endpoints.py:48]: Just use `self.assertIn(str(submission.uuid), self.client.session[SUBMISSIONS_SESSION_KEY])`

File: src/openforms/appointments/tests/test_endpoints.py
Comment (sergei-maertens) [inline src/openforms/appointments/tests/test_endpoints.py:53]: Please use the `furl` dependency (https://pypi.org/project/furl/) that's already in this project to properly mutate/build/parse URLs.

File: src/openforms/appointments/tokens.py
Comment (sergei-maertens) [inline src/openforms/appointments/tokens.py:11]: #655 is still pending, but it has a refactor for token generators in general. Depending on which PR gets merged first, the other will have to adapt to make use of this.

File: src/openforms/appointments/views.py
Comment (sergei-maertens) [inline src/openforms/appointments/views.py:33]: these are not relevant for `RedirectView`

File: src/openforms/appointments/views.py
Comment (sergei-maertens) [inline src/openforms/appointments/views.py:29]: it's not an API endpoint, so we don't document it as such

File: src/openforms/appointments/views.py
Comment (sergei-maertens) [inline src/openforms/appointments/views.py:38]:
```suggestion
        except Submission.DoesNotExist:
```

File: src/openforms/appointments/views.py
Comment (sergei-maertens) [inline src/openforms/appointments/views.py:58]: This is not how you build URLs safely and there is plenty of tooling at your hands to do it the right way:

1. `furl`, mentioned earlier
2. python's stdlib `urlib` package
3. Django's own utils: https://docs.djangoproject.com/en/2.2/ref/utils/#django.utils.http.urlencode

Also - don't go crazy with f-strings, the fact that you can evaluate about anything between the curly braces doesn't mean you should. It should only really be used for drop-in variables, so bring those variables/method calls outside of the f-string.

besides that, doing `f"{str(uuid)}"` is literally the same effect as doing the more readable `f"{uuid}"`, as the string representation will be used by the f-string feature.

File: src/openforms/appointments/views.py
Comment (sergei-maertens) [inline src/openforms/appointments/views.py:54]: This should `raise RunTimeError("No appointment page configured")` if the appointment cancel page URL is empty (the model field is `blank=True` and has no default).

File: src/openforms/emails/templatetags/appointments.py
Comment (sergei-maertens) [inline src/openforms/emails/templatetags/appointments.py:11]: I suspect this will crash if there's no appointment ID in the template context.

**Review by sergei-maertens — CHANGES_REQUESTED** [review]

File: docs/manual/templates.rst
Comment (joeribekker) [reply docs/manual/templates.rst:127]: can we also keep it a little consistant, and rename `get_appointment_links` to `appointment_links` in the template?

**Review by joeribekker — CHANGES_REQUESTED** [review]

File: docs/manual/templates.rst
Comment (sergei-maertens) [reply docs/manual/templates.rst:127]: it's meant to be used as `{% get_appointment_links as links %} {{ links.cancel_url|urlize }}` for example, so I think the `get_` is more semantic. In the future, this will also include the `change_url`.

File: src/openforms/appointments/api/tests/test_endpoints_with_jcc.py
Comment (shea-maykinmedia) [reply src/openforms/appointments/api/tests/test_endpoints_with_jcc.py:287]: The reason was that all the tests require a submission but only some of the tests require the form to be set up in such a way.

That said I can update the tests to use `from_components` now that we have this.

File: src/openforms/appointments/tokens.py
Comment (shea-maykinmedia) [reply src/openforms/appointments/tokens.py:11]: Okay good to know.  I'm guessing you mean once #655 is merged that this will need to inherit from `BaseTokenGenerator`?

File: src/openforms/appointments/api/urls.py
Comment (shea-maykinmedia) [reply src/openforms/appointments/api/urls.py:17]: The reason for using `str` is that the UUIDs of the submissions are stored in the session as strings, so `ActiveSubmissionPermission.has_object_permission` does `submission_id in active_submissions` but if `submission_id` at this point is a UUID then this check will fail and the user would get a 403.

I could update https://github.com/open-formulieren/open-forms/blob/master/src/openforms/submissions/api/permissions.py#L36 to be `return str(submission_id) in active_submissions` if you think that's a better solution?  Or maybe there is a better way solution?

File: src/openforms/emails/templatetags/appointments.py
Comment (shea-maykinmedia) [reply src/openforms/emails/templatetags/appointments.py:11]: It will.  This function requires an appointment id and if there is no appointment id (either because one was not created or because it just wanted added to the context) then this shouldn't be called so I though it was better to crash than fail silently.

File: docs/manual/templates.rst
Comment (shea-maykinmedia) [inline docs/manual/templates.rst:147]: Is this okay for an example?

Screenshot 2021-09-14 at 15 51 19

Screenshot 2021-09-14 at 15 51 25

Comment (joeribekker) [reply docs/manual/templates.rst:147]: Yes, I like

File: src/openforms/emails/templatetags/appointments.py
Comment (sergei-maertens) [reply src/openforms/emails/templatetags/appointments.py:11]: Considering that the content editors/form designers are not experts on the DTL and the forgiveness typically found in the DTL (empty variable -> just output empty string), I think we should give some leeway here. If the context var is not present, just return an empty string.

File: src/openforms/appointments/api/tests/test_endpoints_with_jcc.py
Comment (sergei-maertens) [reply src/openforms/appointments/api/tests/test_endpoints_with_jcc.py]: In any case you want test-data to look realistic, Bart has already been troubled with updating/extending tests that have a very weird, semantically incorrect test setup that then caused problems.

File: src/openforms/appointments/api/urls.py
Comment (sergei-maertens) [reply src/openforms/appointments/api/urls.py:17]: Yes, go ahead with updating the permission class.

And if we were to keep the `str` pattern, _add a comment explaining the non-obvious choice_. We really can't guess from the code itself why the code was written that way, so document design decisions.

File: src/openforms/appointments/tokens.py
Comment (sergei-maertens) [reply src/openforms/appointments/tokens.py:11]: correct, and it's merged now :)

File: src/openforms/appointments/tests/test_base.py
Comment (sergei-maertens) [inline src/openforms/appointments/tests/test_base.py:55]: Remove mocking here and use real token generator

File: src/openforms/appointments/tests/test_base.py
Comment (sergei-maertens) [inline src/openforms/appointments/tests/test_base.py:57]: TODO: set up AppointmentInfo -> is needed for token generation

File: src/openforms/appointments/tests/test_base.py
Comment (sergei-maertens) [inline src/openforms/appointments/tests/test_base.py:71]: use `@override_settings(BASE_URL=...)` instead and verify that the link is indeed the right fully qualified URL.

File: src/openforms/appointments/tests/test_endpoints.py
Comment (sergei-maertens) [inline src/openforms/appointments/tests/test_endpoints.py:18]: use the real token generator instead of mocking

File: src/openforms/appointments/tests/test_endpoints.py
Comment (sergei-maertens) [inline src/openforms/appointments/tests/test_endpoints.py:51]: use string value from earlier set up to make it more explicit - in general you want to avoid "best practices" in tests and keep them as dumb and straight forward as possible.

File: src/openforms/appointments/tests/test_endpoints.py
Comment (sergei-maertens) [inline src/openforms/appointments/tests/test_endpoints.py:65]: Remove mock

File: src/openforms/appointments/tests/test_endpoints.py
Comment (sergei-maertens) [inline src/openforms/appointments/tests/test_endpoints.py:106]: Avoid mocking

File: src/openforms/appointments/tokens.py
Comment (sergei-maertens) [inline src/openforms/appointments/tokens.py:21]: Update docstring to reflect what is actually happening.

File: src/openforms/emails/templatetags/appointments.py
Comment (sergei-maertens) [inline src/openforms/emails/templatetags/appointments.py:15]: `:=` candidate to clean up code a little bit

**Review by sergei-maertens — CHANGES_REQUESTED** [review]
I'll pick up the changes to wrap up this PR, as discussed in the office

**Review by sergei-maertens — APPROVED** [review]

**Pull request merged**
